airbyte_cdk/sources/declarative/migrations/legacy_to_per_partition_state_migration.py

```diff
@@ -47,8 +52,10 @@ def __init__(
             self._cursor.cursor_field, parameters=self._parameters
         ).eval(self._config)
 
-    def _get_partition_field(self, partition_router: SubstreamPartitionRouter) -> str:
-        parent_stream_config = partition_router.parent_stream_configs[0]
+    def _get_partition_field(
+        self, partition_router: CustomPartitionRouter | SubstreamPartitionRouter
+    ) -> str:
+        parent_stream_config = partition_router.parent_stream_configs[0]  # type: ignore # custom partition will introduce this atribute if needed
```

Comment:
[This typing](https://github.com/airbytehq/airbyte-python-cdk/pull/20/files#diff-77adc5cbcd3c74a99502b696f8828068335505eaf9a9342db3218e229bad988dR39) does not make sense to me if we are to ignore there error here. Should we instead fail if the partition router is not of type `SubstreamPartitionRouter` to have a clear error for the dev to take action? Are there cases today where the `partition_router` is not a `SubstreamPartitionRouter`? 

Outside of the scope of this PR: It seems like this wouldn't work in the following cases:
* if it were to be used with `ListPartitionRouter`
* if the `parent_stream_config.partition_field` requires interpolation
* if the first parent is not the "right" parent

Reply:
agreed, reverted to `SubstreamPartitionRouter` for consistency.